src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs

```diff
@@ -58,46 +58,107 @@ private async Task<bool> ShouldShowSpeculativeTCompletionItemAsync(Document docu
                 return false;
             }
 
-            // If we're in a generic type argument context, use the start of the generic type name
-            // as the position for the rest of the context checks.
-            var testPosition = position;
-            var leftToken = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            // We could be in the middle of a ref/generic/tuple type, instead of a simple T case.
+            // If we managed to walk out and get a different SpanStart, we treat it as a slightly different $$T case, otherwise it's a simple T case.
 
-            var semanticModel = await document.GetSemanticModelForNodeAsync(leftToken.Parent, cancellationToken).ConfigureAwait(false);
-            if (syntaxTree.IsGenericTypeArgumentContext(position, leftToken, cancellationToken, semanticModel))
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var semanticModel = await document.GetSemanticModelForNodeAsync(token.Parent, cancellationToken).ConfigureAwait(false);
+
+            var spanStart = position;
+            while (true)
+            {
+                var orgSpanStart = spanStart;
+
+                spanStart = WalkOutOfGenericType(syntaxTree, spanStart, semanticModel, cancellationToken);
+                spanStart = WalkOutOfTupleType(syntaxTree, spanStart, cancellationToken);
+                spanStart = WalkOutOfRefType(syntaxTree, spanStart, cancellationToken);
+
+                if (spanStart == orgSpanStart)
+                {
+                    break;
+                }
+            }
+
+            if (IsStartOfSpeculativeTContext(syntaxTree, spanStart, allowAsyncKeyword: spanStart != position, cancellationToken))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsStartOfSpeculativeTContext(SyntaxTree syntaxTree, int position, bool allowAsyncKeyword, CancellationToken cancellationToken)
+        {
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var prevToken = token.GetPreviousTokenIfTouchingWord(position);
+
+            if (((allowAsyncKeyword || !prevToken.IsKindOrHasMatchingText(SyntaxKind.AsyncKeyword)) &&
```

Comment:
this top line is super hard to read in the context of a massive expresison that contains &&s and ||s.

have tried a few times, and i still keep getting tripped up by it.  can we break this up into well named locals, and not mix/match `&&` `||` `!` etc. :)

Reply:
Are there any plans to support `async T` or we already support it? Since we support `ValueTask<T>` as async method return type.

Reply:
Refactored. I would like to hear your answers on `async T`, if we allow that, these ||s and &&s can be dropped.

Reply:
there isn't a use case i can realistically think of that we'd need to support `async T`.  That said, if it did show up there, i also wouldn't that unhappy.  i.e. it's better for completion to show a little too much.

Reply:
Agree, a `T` in the completion list in such context does no harm. Let's delete all those checks.